Alpha deployment instructions (#741)
corda/corda-docs-portal: content/en/platform/corda/5.0-alpha/deploying/deployment-tutorials/tutorial-k8s.md

---
date: '2021-11-14'

> title: "Deploying a Kubernetes Cluster"

Reviewer: If it is `a` then you probably want `Corda` rather than `Kubernetes
```suggestion
title: "Deploying a Corda cluster"
```

menu:
  corda-5-alpha:
    parent: corda-5-alpha-tutorials-deploy
    identifier: corda-5-alpha-tutorial-deploy-k8s
    weight: 1000
section_menu: corda-5-alpha
---

This page describes how to deploy Corda 5 Alpha. It assumes all necessary [prerequisites](../prerequisites.html) have been installed.

> ## Download and Register Docker Images

Reviewer: I know it's a registry but I've never heard anyone use the verb `register` - `push` is the action in the Docker CLI.
```suggestion
## Download and push Docker images to registry
```

The Corda Docker images must be in a Docker registry that is accessible from the Kubernetes cluster in which Corda will run. The images are provided in a `tar` file which can be loaded into a local Docker engine and then pushed from there to the registry.

1. Download [corda-worker-images.tar]()**add link.

2. Inflate and load the `corda-worker-images.tar` file into the local Docker engine with the following command:
   ```shell
   docker load -i corda-worker-images.tar
   ```

3. Retag each image using the name of the registry to be used and push the image. The following is an example script to automate this. It takes the target Docker registry as an argument.
   ```shell
   #!/bin/bash
   if [ -z "$1" ]; then
    echo "Specify target registry"
    exit 1
   fi

   declare -a images=(
    "corda-os-rpc-worker" "corda-os-flow-worker"
    "corda-os-member-worker" "corda-os-p2p-gateway-worker"
    "corda-os-p2p-link-manager-worker" "corda-os-db-worker"
    "corda-os-crypto-worker" "corda-os-plugins" )

>    tag=5.0.0.0-BetaProgram-alpha.1-RC01

Reviewer: this tag needs to reflect whatever we expect the final version to be

@ronanbrowne - is this going to be `5.0.0.0-BetaProgram-alpha.1`?

Reviewer 2: @davidcurrie apologies just spotted this now.

yes so in this case it will be `5.0.0.0-Eagle` as per yesterday product decision on code names

   target_registry=$1

   for image in "${images[@]}"; do

>     source=corda-os-docker.software.r3.com/$image:$tag

Reviewer: ```suggestion
    source=corda/$image:$tag
```

    target=$target_registry/$image:$tag
    echo "Publishing image $source to $target"
    docker tag $source $target
    docker push $target

>    done

Reviewer: ```suggestion
   done
   
   docker tag postgres:14.4 $target_registry/postgres:14.4
   docker push $target_registry/postgres:14.4
```

```yaml
imagePullSecrets:
  - <IMAGE-PULL-SECRET-NAME>
```

### Replica Counts

For high-availability, specify at least three replicas for each type of Corda worker:
```yaml
workers:
  crypto:
    replicaCount: 3
  db:
    replicaCount: 3
  flow:
    replicaCount: 3
  membership:
    replicaCount: 3
  rpc:
    replicaCount: 3
  p2pGateway:
    replicaCount: 3
  p2pLinkManager:
    replicaCount: 3
```

{{< note >}}
Depending on your application workload, you may require additional replicas.
{{< /note >}}

### Resource Requests and Limits

Specify a default set of resource requests and limits for the Corda containers:
```yaml
resources:
  requests:
    memory: 512Mi
    cpu: 125m
  limits:
    memory: 2048Mi
    cpu: 1000m
```
{{< note >}}
It is particularly important to specify resource requests when using a Kubernetes cluster with auto-scaling to ensure that it scales appropriately when the Corda cluster is deployed.
{{< /note >}}

You can also override the default resource requests and limits separately for each type of Corda worker. For example, to increase the memory limit for flow workers:
```yaml

> flow:
>   resource:
>     limit:
>       memory: 4096Mi

Reviewer: My mistake...
```suggestion
workers:
  flow:
    resource:
      limit:
        memory: 4096Mi
```

```
As with the number of replicas, you may need to adjust these values based on testing with your actual application workload.

### REST API Load Balancer

By default, the [REST API](../../developing/rest-api/rest-api.html) is exposed on an internal Kubernetes service. To enable access from outside the Kubernetes cluster, the API should be fronted by a load balancer. The Helm chart allows annotations to be specified to facilitate the creation of a load balancer by a cloud-platform specific controller. For example, the following configuration specifies that the [AWS Load Balancer Controller](https://docs.aws.amazon.com/eks/latest/userguide/aws-load-balancer-controller.html) fronts the REST API with a Network Load Balancer internal to the Virtual Private Cloud (VPC):
```yaml
workers:
  rpc:
    service:
      type: LoadBalancer
      annotations:
        service.beta.kubernetes.io/aws-load-balancer-internal: true
        service.beta.kubernetes.io/aws-load-balancer-scheme: internal
        service.beta.kubernetes.io/aws-load-balancer-type: nlb

>         external-dns.alpha.kubernetes.io/hostname: rpc-worker.dev1.corda.cloud

Reviewer: Changing to something that isn't R3 internal...
```suggestion
        external-dns.alpha.kubernetes.io/hostname: corda.example.com
```

```

### PostgreSQL

> The password for PostgreSQL can be specified directly as a Helm override but this is not recommended. Instead, create a Kubernetes secret containing the password with a key of password. By default, the install expects a database called `cordacluster` but this can be overridden. You can then define the PostgreSQL configuration as follows:

Reviewer: ```suggestion
The password for PostgreSQL can be specified directly as a Helm override but this is not recommended. Instead, create a Kubernetes secret containing the password with a key of `password`. By default, the install expects a database called `cordacluster` but this can be overridden. You can then define the PostgreSQL configuration as follows:
```

```yaml
db:
  cluster:

>     host: <POSTGRESQL_HOST>
>     port: 5432
>     user: <POSTGRESQL_USER>
>     database: <POSTGRESQL_CLUSTER_DB>
>     existingSecret: <POSTGRESQL_PASSWORD_SECRET_NAME>

Reviewer: ```suggestion
    host: <POSTGRESQL-HOST>
    port: 5432
    database: <POSTGRESQL-CLUSTER-DB>
    user:
      value: <POSTGRESQL-USER>
    password:
      valueFrom:
        secretKeyRef:
          name: <POSTGRESQL-PASSWORD-SECRET-NAME>
          key: "password"
```

```

### Kafka

Specify the Kafka bootstrap servers as a comma-separated list:
```yaml
kafka:

>   boostrapServers: <KAFKA_BOOTSTRAP_SERVERS>

Reviewer: ```suggestion
  boostrapServers: <KAFKA-BOOTSTRAP-SERVERS>
```

```
If desired, a prefix can be applied to all of the Kafka topics used by the Corda deployment. This enables multiple Corda clusters to share a Kafka cluster. For example:
```yaml
kafka:

>   topicPrefix: <KAFKA_TOPIC_PREFIX>

Reviewer: ```suggestion
  topicPrefix: <KAFKA-TOPIC-PREFIX>
```

```
Enable TLS if required by the Kafka client protocol:
```yaml
kafka:
  tls:
    enabled: true
```

If the broker certificate is self-signed or can not be trusted for some other reason, create a Kubernetes secret containing the client trust store. The trust store can be in PEM or JKS format. If JKS format is used, you can supply a password for the trust store. The following example is for a trust store in PEM format stored against the `ca.crt` key in the Kubernetes secret:
```yaml
kafka

>   tls:
>     secretRef:
>       name: <TRUST-STORE-SECRET-NAME>
>       key: "ca.crt"
>     type: PEM   

Reviewer: ```suggestion
  tls:
    truststore:
      valueFrom:
        secretKeyRef:
          name: <TRUSTSTORE-SECRET-NAME>
          key: "ca.crt"
    type: PEM  
```

```

> Corda supports SASL for Kafka authentication. If your Kafka instance requires SASL authentication, enable this and specify the necessary credentials along with the mechanism to be used:

Reviewer: ```suggestion
Corda supports SASL for Kafka authentication. If your Kafka instance requires SASL authentication, create a secret containing the credentials with the keys username and password and then specify the secret name, along with the required mechanism in the overrides:
```

```yaml
kafka:
  sasl:
    enabled: true

>     username: <KAFKA_USERNAME>
>     password: <KAFKA_PASSWORD>

Reviewer: ```suggestion
    username:
      valueFrom:
        secretKeyRef:
          name: <SASL-SECRET-NAME>
          key: "username"
    password:
      valueFrom:
        secretKeyRef:
          name: <SASL-SECRET-NAME>
          key: "password"
```

You may need to increase this to support a larger number of Corda worker replicas.
It is also possible to override the default topic replica count.
For example, if less than three Kafka brokers are available.
The following extract shows the default values:

```yaml
bootstrap:
  kafka:
    enabled: true
    partitions: 10
    replicas: 3
```

#### Database

The database bootstrapping creates schemas in the cluster database and populates them with the initial configuration. It is enabled by default:

```yaml
bootstrap:
  db:
    enabled: true
```
By default, the database bootstrapping uses the psql CLI from the Docker image `postgres:14.4` on Docker Hub. If the Kubernetes cluster does not have access to Docker Hub, this image must be made available in an internal registry. The location of the image can then be specified via overrides, as follows:

```yaml
db:
  clientImage:
    registry: <REGISTRY>
    repository: "postgres"
    tag: "14.4"
```
Part of the database bootstrapping involves populating the initial admin credentials. You can specify these in one of the following ways:

* Pass the username and password as Helm values:
  ```yaml

>   bootstrap:
>   initialAdminUser:
>     username: <USERNAME>
>     password: <PASSWORD>

Reviewer: ```suggestion
bootstrap:
  initialAdminUser:
    username:
      value: <USERNAME>
    password:
      value: <PASSWORD>
```

  ```

* If the password field is left blank, a Kubernetes secret is created containing a generated password. The notes output when the deployment completes contain instructions for how to retrieve this. This is the default behavior.

> * Create a Kubernetes secret containing the user credentials. By default, Kubernetes expects the secret to contain the username and password of the key:

Reviewer: ```suggestion
* Create a Kubernetes secret containing the user credentials:
```

  ```yaml

>   bootstrap:
>   initialAdminUser:
>     secretRef:
>       name: <INITIAL-ADMIN-USER-SECRET-NAME>
>       usernameKey: "username"
>       passwordKey: "password"

Reviewer: ```suggestion
bootstrap:
  initialAdminUser:
    username:
      valueFrom:
        secretKeyRef:
          name: <INITIAL_ADMIN_USER_SECRET_NAME>
          key: "username"
    password:
      valueFrom:
        secretKeyRef:
          name: <INITIAL_ADMIN_USER_SECRET_NAME>
          key: "password"
```

---
date: '2021-11-14'

> title: "Deploying a Corda Cluster"

Reviewer 3: If the title is "Deploying a Corda Cluster" - shouldn't the file be renamed as well ?

menu:
  corda-5-alpha:
    parent: corda-5-alpha-tutorials-deploy
    identifier: corda-5-alpha-tutorial-deploy-k8s
    weight: 1000
section_menu: corda-5-alpha
---

This page describes how to deploy Corda 5 Alpha. It assumes all necessary [prerequisites](../prerequisites.html) have been installed.

## Download and Push Docker Images to a Registry

The Corda Docker images must be in a Docker registry that is accessible from the Kubernetes cluster in which Corda will run. The images are provided in a `tar` file which can be loaded into a local Docker engine and then pushed from there to the registry.

> 1. Download `corda-worker-images.tar` from the [R3 Customer Hub](https://r3.force.com/).

Reviewer 3: looking at what @ronanbrowne created, `corda-worker-images.tar` will probably be suffixed w/ version, i.e. `corda-worker-images-<version>.tar`

Author: Changed to `corda-worker-images-eagle.tar`
